### PR TITLE
feat: Support row virtualization using react-window in table widget

### LINKED ISSUE #16567: [Task]: Row virtualization in Table widget
### Is there an existing issue for this?

- [X] I have searched the existing issues

### SubTasks

Leverage react table virtualized row feature to improve performance in the table widget.

### LINKED ISSUE #17178: [Task]: Table widget performance optimisation - part 1
### Is there an existing issue for this?

- [X] I have searched the existing issues

### SubTasks

 - [ ] Refactor autoToolTip component calculation to  improve the performance.

### PR DESCRIPTION
**Pull request by sbalaji1192**
Introducing the virtual row functionality to improve the performance of the table when there are a lot of rows to render.
React-window library is used to implement the virtual row functionality. This functionality will only be enabled if the cell wrapping of all the columns is disabled since the virtual expects each row of constant height.

Fixes #16567
Fixes #17178

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
Following cypress tests have been added 
1. should check that row is getting rendered.
2. should check that virtual rows are getting rendered when scrolling through the table
3. should check that the virtual rows feature is turned off when cell wrapping is enabled

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by laveena-en** [comment]
Tested and verified for:
Row Virtualization
Regression on the table widget

File: app/client/src/widgets/TableWidgetV2/component/Table.tsx [hunk 1]
```diff
@@ -360,67 +377,25 @@ export function Table(props: TableProps) {
...
-                  props.borderRadius,
-                )}
-            </div>
+              rows={subPage}
+              selectTableRow={props.selectTableRow}
+              selectedRowIndex={props.selectedRowIndex}
+              selectedRowIndices={props.selectedRowIndices}
+              tableSizes={tableSizes}
+              totalColumnWidth={totalColumnWidth}
+              useVirtual={!isCellWrappingAllowed}
```
Comment (ashit-rath) [inline app/client/src/widgets/TableWidgetV2/component/Table.tsx:396]: So we do not allow virtualization when any one of the columns have cell wrapping enabled?
Also should we enable only when it is server side pagination? as the client side one would always have fixed rows

File: app/client/src/widgets/TableWidgetV2/component/Table.tsx [hunk 2]
```diff
@@ -228,6 +227,24 @@ export function Table(props: TableProps) {
...
+    return props.columns.reduce((prev, curr) => {
+      return prev + (curr.width || DEFAULT_COLUMN_WIDTH);
+    }, offset);
+  }, [columnString, props.multiRowSelection]);
+
+  const isCellWrappingAllowed = useMemo(() => {
+    return props.columns.some(
+      (column) => column.columnProperties.allowCellWrapping,
+    );
+  }, [columnString]);
```
Comment (ashit-rath) [inline app/client/src/widgets/TableWidgetV2/component/Table.tsx:246]: I don't think we need to memoise this as `isCellWrappingAllowed` is a primitive value and when it's passed to `Row` react can handle with shallow equals.

File: app/client/src/widgets/TableWidgetV2/component/Table.tsx [hunk 3]
```diff
@@ -228,6 +227,24 @@ export function Table(props: TableProps) {
     [props.width],
   );
 
+  const totalColumnWidth = useMemo(() => {
```
Comment (ashit-rath) [inline app/client/src/widgets/TableWidgetV2/component/Table.tsx:230]: Same here, no need to memoise. The computation doesn't seem to be heavy and the return value is a primitive.

**Review by ashit-rath — COMMENTED** [review]
@sbalaji1192 Can you update the PR description?

File: app/client/src/widgets/TableWidgetV2/component/Row.tsx [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+  const isRowSelected = props.multiRowSelection
+    ? props.selectedRowIndices.includes(props.row.index)
+    : props.row.index === props.selectedRowIndex;
+  return (
+    <div
+      {...rowProps}
+      className={"tr" + `${isRowSelected ? " selected-row" : ""}`}
+      data-rowindex={props.index}
+      key={props.index}
+      onClick={(e) => {
```
[the remark this answers is not in the document]
Comment (sbalaji1192) [reply app/client/src/widgets/TableWidgetV2/component/Row.tsx:47]: This code is already part of an existing component and I'm just extracting it to create a row component. We're not adding accessibility for now.

File: app/client/src/widgets/TableWidgetV2/component/Row.tsx [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+  (props: RowPropsType, ref: Ref<HTMLDivElement>) => {
+    return (
+      <div {...props.getTableBodyProps()} className="tbody" ref={ref}>
+        {props.rows.map((row, index) => {
+          return (
+            <BasicRow
+              accentColor={props.accentColor}
+              borderRadius={props.borderRadius}
+              index={index}
+              key={index}
```
[the remark this answers is not in the document]
Comment (sbalaji1192) [reply app/client/src/widgets/TableWidgetV2/component/Row.tsx:157]: No. It's a plain object.

File: app/client/src/widgets/TableWidgetV2/component/Table.tsx (hunk 1 above)
Comment (sbalaji1192) [reply app/client/src/widgets/TableWidgetV2/component/Table.tsx:396]: Yeah. good point. Let me change the condition here.

File: app/client/src/widgets/TableWidgetV2/component/Table.tsx (hunk 2 above)
Comment (sbalaji1192) [reply app/client/src/widgets/TableWidgetV2/component/Table.tsx:246]: We are memoizing it to avoid repeated calculations.

Comment (ashit-rath) [reply app/client/src/widgets/TableWidgetV2/component/Table.tsx:246]: I don't think the calculations are very expensive as the number of columns cannot be that very high. 
Also the `some` does early exit if it encounters a `true`

Comment (sbalaji1192) [reply app/client/src/widgets/TableWidgetV2/component/Table.tsx:246]: I have removed it.

File: app/client/src/widgets/TableWidgetV2/component/Table.tsx [hunk 6]
```diff
@@ -228,6 +225,11 @@ export function Table(props: TableProps) {
     [props.width],
   );
 
+  const shouldUseVirtual =
+    !props.columns.some(
+      (column) => !!column.columnProperties.allowCellWrapping,
+    ) && props.serverSidePaginationEnabled;
```
Comment (ashit-rath) [inline app/client/src/widgets/TableWidgetV2/component/Table.tsx:231]: Can we have `props.serverSidePaginationEnabled` as the first condition and then the `columns.some`? 
If sever side is disabled then the 2nd condition doesn't need to run.

File: app/client/src/widgets/TableWidgetV2/component/cellComponents/AutoToolTipComponent.tsx [hunk 7]
```diff
@@ -39,14 +39,13 @@ function useToolTip(
   const [showTooltip, updateToolTip] = useState(false);
 
   useEffect(() => {
-    requestAnimationFrame(() => {
```
Comment (ashit-rath) [inline app/client/src/widgets/TableWidgetV2/component/cellComponents/AutoToolTipComponent.tsx:42]: Why did we remove `requestAnimationFrame`?

File: app/client/src/widgets/TableWidgetV2/component/cellComponents/EmptyCell.tsx [hunk 8]
```diff
@@ -41,26 +45,18 @@ export const renderEmptyRows = (
...
-                );
-              })}
-            </EmptyRow>
-          );
-        })}
-      </>
-    );
+    return rows.map((row: string, index: number) => {
+      return (
+        <EmptyRow className="tr" key={index} style={style}>
```
Comment (ashit-rath) [inline app/client/src/widgets/TableWidgetV2/component/cellComponents/EmptyCell.tsx:50]: Is it possible to use `primary column` value for the `key` here? and if it's not present then as a fallback the `index`?

File: app/client/src/widgets/TableWidgetV2/component/cellComponents/AutoToolTipComponent.tsx (hunk 7 above)
Comment (sbalaji1192) [reply app/client/src/widgets/TableWidgetV2/component/cellComponents/AutoToolTipComponent.tsx:42]: requestAnimationFrame is creating an opposite effect here. its triggering a layout thrashing for each cell in the Table. hence we removed it.
Screenshot 2022-09-29 at 10 47 43 AM

File: app/client/src/widgets/TableWidgetV2/component/cellComponents/EmptyCell.tsx (hunk 8 above)
Comment (sbalaji1192) [reply app/client/src/widgets/TableWidgetV2/component/cellComponents/EmptyCell.tsx:50]: This is an empty row we render to fill the table body. it doesn't have any data.

**Review by ashit-rath — DISMISSED** [review]

**Comment by laveena-en** [comment]
Tested and verified:

-Performance testing with large number of records (1000 records, 40 cols)
-Regression on table widget.
-Functionality in view and edit mode.

**Review by ashit-rath — APPROVED** [review]

**Pull request merged**